R/utilities-simulation.R

```diff
@@ -134,37 +134,68 @@ simulateModelOnCore <- function(simulation,
   return(simulationResult)
 }
 
+
+#' @title defaultMaxSimulationsPerCore
+#' @description Default value for a scale factor used in a parallel simulation.  The product of this scale factor and the number of allowable cores (allowedCores) sets the maximum number of simulations that may be run on one core.
+#' @export
+defaultMaxSimulationsPerCore <- 2
+
+
+
 #' @title simulateModelParallel
 #' @description Simulate models within a list of structure sets in parallel for an individual.
 #' @param structureSets, a list of `SimulationStructure` R6 class objects contain paths of files to be used
 #' @param settings list of options to be passed to the function
 #' @param logFolder folder where the logs are saved
 #' @return List of simulation results for each simulation set
 #' @import ospsuite
+#' @import parallel
 #' @keywords internal
 simulateModelParallel <- function(structureSets,
                                   settings = NULL,
                                   logFolder = getwd()) {
-  simulations <- lapply(structureSets, function(set) {
-    re.tStoreFileMetadata(access = "read", filePath = set$simulationSet$simulationFile)
-    simulation <- loadSimulationWithUpdatedPaths(set$simulationSet)
+
+  simulationResults <- list()
+  maxSimulationsPerSubset <- settings$maxSimulationsPerCore*((settings$allowedCores %||% parallel::detectCores()) - 1)  #To be set in settings argument
+

```

Comment:
@Yuri05 since you touched on this in #347 :
- is `settings$allowedCores` redundant here?  ie does `parallel::detectCores()` work on linux as in windows?
- does `settings$allowedCores` return  `parallel::detectCores()` or does it return `parallel::detectCores() - 1`

Reply:
Why do we have settings$allowedCores ? Isn't is somehow redundant with the cores one? And how can it be undefined then?

Reply:
OK `getOSPSuiteSetting` should get rid of the redundancy here.  But, indeed, what is the need for `getAllowedCores` (assuming `parallel::detectCores()` is fine in linux).

Reply:
`getAllowedCores()` will return a value only if RE is executed on a **Kubernetes Cluster on Linux** (to assure that RE does not take more cores than are assigned to the user in the cluster session). In all other cases it will return NULL.

So to get the number of cores we must do something like this:
```
numberOfCores = getAllowedCores() 
if (numberOfCores == NULL)
{
    numberOfCores = getOSPSuiteSetting("numberOfCores")
}

```

Reply:
I think that should be encapsulated in a function. And maybe getAllowedCore should be renamed so that we know that it's only a specific case?

Reply:
>I think that should be encapsulated in a function.

agree

>And maybe getAllowedCore should be renamed so that we know that it's only a specific case?

Yes. But don't ask me for the new function name 😄